### ISSUE TITLE
Narrow sometimes lands far in the past

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-mobile#3457]
A user reports:

>  if Zulip Chat is open and a new message is incoming, that it not direct scroll to the message if you tab it in the unread message page, the message is not there. If i close, open Zulip, it works.
Similar is if you open any topic it won't scroll to bottom to newest message, you have do it manually.

When asked

> Is the message there if you scroll down to it, or is the message not present even when you scroll?

they add:

> Message is present, but need to scroll down multiple times until bottom to reach the new message. On each scroll step i see the arrow down icon to jump to the bottom. If press that icon it just jumps som already loaded messages down and so on.... after repeat scroll down multiple times, bottom is reached.

From further discussion I think the scenario is that there's just one (or maybe a handful) of unread messages, in a conversation with many older messages; and the narrow lands far in the past, many messages before the unreads.

They're using v23.3.112 on Android 9.

The symptom is reminiscent of some ancient bugs: #1545, #2146 (minus the mark-as-read part), #2699. The latest of those was fixed last June, though, and I don't think we've seen such symptoms since then, so the cause is likely unrelated to those.

**Comment by gnprice** [comment]
A user just reported a pattern that might be a description of this issue:

> Occasionally conversations "autoscroll" to the top (or at least quite far back) in a topic, requiring manually scrolling back down

I'm not sure whether they're referring to what they first see when narrowing to the conversation, which is what would make it the same as this issue. It kind of sounds like they mean they're already looking at the messages when they get scrolled way up, which would be a different set of symptoms.

**Comment by gnprice** [comment]
User feedback today (on a related [chat thread](https://chat.zulip.org/#narrow/stream/48-mobile/topic/start.20position.20when.20reading.20topics/near/746531)):
> BTW, I still get this almost every time I use the app and I think it contributes to a perception that the Zulip app is broken and janky

I've also found in my own use that this is one of the top annoyances I've noticed in recent weeks.

In my experience it's often just 1-2 screenfuls back.

**Comment by gnprice** [comment]
I just tried a few times to reproduce this, narrowing to various streams and PM conversations on chat.zulip.org. I wasn't able to do so.

So one very useful form of progress on this would be to experiment more, and find a way to reproduce the issue at will.

**Comment by gnprice** [comment]
Some user reports from that same [chat thread](https://chat.zulip.org/#narrow/stream/48-mobile/topic/start.20position.20when.20reading.20topics/near/746531):

> I just tried loading many different narrows of different types (stream, topic, all messages, PMs), and I think every one of them other than "All messages" started (on the first load) at or near the first message in the narrow; "All messages" started at a more recent message. This is on a month-old server, so maybe there's a number of messages back that the app ends up starting at, which for most narrows was enough to start at the first message.
>
> After going back then reloading each narrow, all except "All messages" were at the bottom, except for one that was scrolled up just enough to hide the content of the last message (but not the sender's name).

Another person:

> So I’ve figured it out. The pointer doesn’t update when you first go into a stream but refreshes if you focus on a stream, back out, then go back into it again.
> So that’s the way around it I’ve found for now

See also some investigation by @ishammahajan on #3492 .

**Comment by gnprice** [comment]
> See also some investigation by @ishammahajan on #3492 .

In particular:

> I have noticed that the scroll position is often misplaced when opening streams which haven't been cached (not been opened before). More specifically, the scroll position seems to be on the top instead of the bottom of the list -- and consequently continues to load more messages since it thinks we scrolled to the top of the list.
> 
> This only reproduces on `yarn ios-min` though (not on android, emulator or device. Not on ios-max),

**Comment by ray-kraesig** [comment]
Recently I ran across some unusually-consistently-inconsistent behavior that seems like it could be at least closely related to this.

The test case involved repeatedly entering and leaving a stream (with all messages read) containing a handful of unusually-long individual messages. (First lines of Edwardian novels.) This was done on an emulated Android device, connecting to a dev server running on the same physical machine. No reloading of the app or its assets was performed.

When `handleInitialLoad` is called, `document.body.scrollHeight` and `.clientHeight` would take on one of two pairs of values:

* `.scrollHeight` = 3730, `.clientHeight` = 607
* `.scrollHeight` = 8249, `.clientHeight` = 0

The latter, of course, is incorrect; its appearance was consistently associated with the message-list initially being scrolled to a point in the past. Conversely, the former was consistently associated with the message-list initially being scrolled to the bottom, as expected.

----

What brought this to my attention was the fact that, in an intermediate commit, I had mistakenly been loading a large chunk (240K+) of unnecessary JavaScript in the `<head>` of the message-list HTML. While this chunk was in place, the correct values were always seen in `handleInitialLoad`, and the message-list would always appear at the bottom.

Presumably, this chunk was delaying JavaScript execution long enough for the first layout/reflow to occur.  Pushing the scroll-control-function calls found in `handleInitialLoad` into a `requestAnimationFrame` callback seemed to fix the issue, as well. 

This last is _probably_ the right basic approach, but may have to be done in multiple locations and/or in a more principled fashion.

**Comment by gnprice** [comment]
A very interesting lead! That could well be the cause of this bug.

Reading up on `requestAnimationFrame`:
* https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame
* https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Performance_best_practices_for_Firefox_fe_engineers (for a slightly different context but most of the discussion should apply)

it sounds like it's not the right solution in principle: it calls the callback just before the style and layout phases, and we want to run just after. In particular, querying for something like `scrollHeight` inside a `requestAnimationFrame` callback is likely to cause "synchronous reflow"... which IIUC is likely to mean a dropped frame.

Ah, but [here's](https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Performance_best_practices_for_Firefox_fe_engineers#How_do_I_avoid_triggering_uninterruptible_reflow) a solution in that latter page that is applicable:
> It's also possible (though less infallible than [a non-web, Firefox-internal method]) to queue JavaScript to run very soon after the frame has been painted, where the likelihood is highest that the DOM has not been written to, and layout and style information queries are still cheap.
```
requestAnimationFrame(() => {
  setTimeout(() => {
    // This code will be run ASAP after Style and Layout information have
    // been calculated and the paint has occurred. Unless something else
    // has dirtied the DOM very early, querying for style and layout information
    // here should be cheap.
  }, 0);
});
```

That might be a good solution.

**Comment by ray-kraesig** [comment]
> In particular, querying for something like `scrollHeight` inside a `requestAnimationFrame` callback is likely to cause "synchronous reflow"... which IIUC is likely to mean a dropped frame.

That sounds like exactly what we want, though. (At least in `handleInitialLoad`. Notsomuch elsewhere.) If the first paint takes place before we change the scroll location, the first frame will show the top of the message-list page, immediately followed by a flicker to the bottom. _That_ frame, I think, we'd rather drop.

On the other hand, I'm a bit puzzled as to why we're not already causing a reflow (as evidenced by the incorrect values) when we access `scrollHeight` in the current code.

**Comment by gnprice** [comment]
> That sounds like exactly what we want, though. (At least in `handleInitialLoad`. Notsomuch elsewhere.) If the first paint takes place before we change the scroll location, the first frame will show the top of the message-list page, immediately followed by a flicker to the bottom. _That_ frame, I think, we'd rather drop.

Hmm yeah, maybe so.



> On the other hand, I'm a bit puzzled as to why we're not already causing a reflow (as evidenced by the incorrect values) when we access `scrollHeight` in the current code.

Yeah, seems like there must be something else that hasn't happened yet when we get a wrong value there.

The different `scrollHeight` values seem like a clue. Do you have a hypothesis for what is making the `scrollHeight` so big initially? Perhaps a bunch of elements that should be hidden are getting included in the height; could be interesting to look at what's supposed to hide them.

**Comment by gnprice** [comment]
(More discussion today [in chat](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/clientHeight.20and.20friends/near/798730).)

**Comment by gnprice** [comment]
OK, here are some things we learned in the chat discussion:
* Both in @ray-kraesig 's repro, and when I repro it (both of us in Android emulators), the symptom is we land right about one screenful too high. (To within <2px in Ray's case; about 54px more than a screenful in mine.)
* The observations above, where `scrollHeight` is too big, are kind of puzzling as a cause for that symptom! Because in that case `scrollToAnchor` should be trying to scroll us too far *down*.
* But, we also found a diagnosis for something that puts us a screenful too high! The `resize` handler is getting called as we go from a `clientHeight` of 0 to the actual `clientHeight`. When the viewport gets taller, that handler is meant to scroll us up (as measured from the top of the viewport to the top of the content) in order to keep the bottom of the viewport at the same spot in the content. So, it dutifully scrolls us up by one screenful.
* The following patch, which cuts out that spurious resize, resolves the issue completely in my repro case:
```diff
 let viewportHeight = documentBody.clientHeight;
 
 window.addEventListener('resize', event => {
+  if (viewportHeight === 0) {
+    viewportHeight = documentBody.clientHeight;
+    return;
+  }
   const difference = viewportHeight - documentBody.clientHeight;
```
* I think in fact that should be a 100% fix for the case where the desired location is at the bottom, i.e. the caught-up case. The initial `scrollToAnchor` will try to scroll too far down; we'll get clamped to the actual bottom, and all's well. But if we meant to show the user something in the middle, then scrolling too far down is already bad.
* So I think we need to *both*
  * delay the initial `scrollToAnchor` until a point where we have sensible heights, as discussed above; and
  * avoid spurious scrolling on a `resize` from zero.
* More generally, we shouldn't use `viewportHeight` when it has a nonsense value. (Are other nonsense values possible besides zero?) I looked at other places where we use `viewportHeight`, and both of them will make a reasonable approximation of doing nothing if `viewportHeight` is zero:
  * `visibleMessageIds` will return an empty range.
  * `findPreserveTarget` will return `{ type: 'none' }`.

  (Both will call `someVisibleMessage(0, 0)`, which will return null.)

  It'd be good to make that more direct and explicit, though.

---

I think at this point the main open question may be: what's an appropriate way to wait for the heights to get reasonable values set? Some discussion above in this thread; I don't think we expanded much on that point in the chat thread.

Oh, except I linked to this bit of context that Ray had shown me the other day, [in a spec](https://www.w3.org/TR/cssom-view/#dom-element-clientheight):
> If the element has no associated CSS layout box or if the CSS layout box is inline, return zero.

Where following a link there finds that the term "CSS layout box" is not actually defined. :disappointed: And that it's been that way since at least 2008, so it's not like this is a recently-drafted fast-moving spec that may get that point filled in any day now.

Anyway, though, that at least explains why zero in particular is a nonsense value we might see -- and offers encouragement that that might be the only nonsense case.

There were also some helpful pointers from Tim into the webapp code, which I haven't yet fully studied. That could be a very helpful comparison.

**Comment by gnprice** [comment]
> Recently I ran across some unusually-consistently-inconsistent behavior that seems like it could be at least closely related to this. [...]

[This comment](https://github.com/zulip/zulip-mobile/issues/3457#issuecomment-555234717) and the several after it in this thread, up to just now, describe an issue that @ray-kraesig discovered which causes the message list to be scrolled one screenful too far back. I've just filed #3921 for that issue, copying there what we learned about it in this thread. Let's continue discussion of that issue there.

The reason I've split it off is: we have a [fresh report](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/scroll.20position/near/818819) of the kind of symptoms described in the early reports in this thread. With a [reproduction recipe](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/scroll.20position/near/819145)! And confirmation that the amount of scrollback can be many screenfuls, which shows it's a distinct issue from the one that's now #3921.

I think the reproduction recipe means we're much closer to being able to resolve this issue than we have been before.

**Comment by chrisbobbe** [comment]
@gnprice [said](https://github.com/zulip/zulip-mobile/issues/3921#issuecomment-601464395):
> > This, in turn, could be done by changing `getFirstUnreadIdInNarrow` to return the last message in the list if there are no unreads.
> 
> Hmm! Yeah, I agree -- this sounds like a promising solution. As we discussed the other day, it'd be good to have `getFirstUnreadIdInNarrow` (or some other code that's still in the main JS environment with all our data) compute that, so that we can always pass a specific message ID in for the webview JS code to scroll to.

This would basically be a diff like

```diff
 export const getFirstUnreadIdInNarrow: Selector<number | null, Narrow> = createSelector(
   (state, narrow) => getShownMessagesForNarrow(state, narrow),
   getFlags,
   getSubscriptions,
   getMute,
   (messages, flags, subscriptions, mute) => {
-    const firstUnread = messages.find(msg => !flags.read[msg.id]);
+    const firstUnread = messages.find(msg => !flags.read[msg.id]) ?? messages[messages.length - 1];
     return firstUnread?.id ?? null;
   },
 );
```

and all the variable renaming and test changes that would need to follow.

**Comment by gnprice** [comment]
We've had a couple of fresh reports of this [in chat](https://chat.zulip.org/#narrow/stream/48-mobile/topic/Wrong.20scroll.20position.20on.20iOS/near/1084864). One person says (and adds they're on iOS):
> Almost every time I access a stream (either from a notification or just from the app, PM or channel) the app places me 10-20 (haven't been able to work out the distance exactly) messages back up the history even though it's all read. This has been happening for 6+ months so I figure it's a long running bug that I'm only just getting around to reporting.

On my asking whether it's about a screenful or more than that:
> Yep […] it's certainly more than a screenshot worth. I just did a quick measurement and it was approx 14 screens of conversation.

**Comment by gnprice** [comment]
OK, and on that chat thread we have a new and simpler repro recipe!

Moreover it seems to be specific to iOS -- the same recipe does not reproduce it for me on Android.

I'll file a new issue for this one; the original report at the top of this thread was on Android, so is probably not the same issue, and this issue thread has also gotten confusing with what are probably several different issues (with symptoms that make them subtle to distinguish from each other), so a new issue will help focus the debugging

**Comment by gnprice** [comment]
Now that #4357 has been diagnosed and fixed :tada:, I think that may be the resolution of this issue!

I don't recall seeing this issue myself, as a mostly-Android user, in a long time. I suspect that the situation for some time has been that the issue existed only on iOS, and #4357 was it.

I've just closed #3921, which was forked from this issue, because I can't reproduce it anymore either.

Closing this issue too. If we see symptoms like these again in the future, let's file a new bug and investigate.

There's still a potential opportunity here to simplify our interactions with the browser's layout and scrolling:
https://github.com/zulip/zulip-mobile/issues/3457#issuecomment-613095764
That could potentially make us more robust to having an issue like this appear in the future. If we see another issue like this one in the future, that's one direction we'll want to investigate.
